core/shared/src/main/scala/scalaz/zio/IO.scala

```diff
@@ -358,6 +369,18 @@ sealed abstract class IO[+E, +A] extends Serializable { self =>
         }
     )(_ => self)
 
+  /**
+   * Runs the specified action if this action is interrupted.
+   */
+  final def onInterrupt(cleanup: ExitResult[E, Nothing] => IO[Nothing, Unit]): IO[E, A] =
+    IO.bracket0(IO.unit)(
```

Comment:
Unfortunately (in some respects) `redeem0` is ignored for interruption. I think the interruption flag may be more promising. Because this is starting to be a huge amount of overhead to do something on interruption.